packages/hardhat-ledger/src/provider.ts

```diff
@@ -0,0 +1,445 @@
+
+    let path = "";
+    try {
+      for (
+        let account = 0;
+        account <= LedgerProvider.MAX_DERIVATION_ACCOUNTS;
+        account++
+      ) {
+        path = `44'/60'/${account}'/0/0`;
+
+        this.emit("derivation_progress", path, account);
+
+        const wallet = await this.eth.getAddress(path);
+        const address = wallet.address.toLowerCase();
+
+        if (address === addressToFind) {
+          this.emit("derivation_success", path);
+          this.paths[addressToFind] = path;
+
+          void cache.write(this.paths); // hanging promise
+
+          return path;
+        }
+      }
+    } catch (error) {
+      const message = (error as Error).message;
+
+      this.emit("derivation_failure");
+      throw new HardhatLedgerDerivationPathError(
+        `There was an error trying to derivate path ${path}: "${message}". The wallet might be connected but locked or in the wrong app.`,
+        path
+      );
+    }
+
+    this.emit("derivation_failure");
+    throw new HardhatLedgerDerivationPathError(
+      `Could not find a valid derivation path for ${addressToFind}. Paths from m/44'/60'/0'/0/0 to m/44'/60'/${LedgerProvider.MAX_DERIVATION_ACCOUNTS}'/0/0 were searched.`,
+      path
+    );
+  }
+
+  private async _withConfirmation<T extends (...args: any) => any>(
+    func: T
+  ): Promise<ReturnType<T>> {
+    try {
+      this.emit("confirmation_start");
+      const result = await func();
+      this.emit("confirmation_success");
+
+      return result;
+    } catch (error) {
+      this.emit("confirmation_failure");
+      throw new HardhatLedgerError((error as Error).message);
+    }
+  }
+
+  private async _toRpcSig(signature: Signature): Promise<string> {
+    const { toRpcSig, toBuffer } = await import(
+      "@nomicfoundation/ethereumjs-util"
+    );
+
+    return toRpcSig(
+      BigInt(signature.v - 27),
+      toBuffer(this._toHex(signature.r)),
+      toBuffer(this._toHex(signature.s))
+    );
+  }
+
+  private async _getNonce(address: Buffer): Promise<bigint> {
+    const { bufferToHex } = await import("@nomicfoundation/ethereumjs-util");
+
+    const response = (await this._wrappedProvider.request({
+      method: "eth_getTransactionCount",
+      params: [bufferToHex(address), "pending"],
+    })) as string;
+
+    return rpcQuantityToBigInt(response);
+  }
+
+  private async _requireControlledInit(address: Buffer): Promise<void> {
+    this._requireControlledAddress(address);
+    await this.init();
+  }
+
+  private _requireControlledAddress(address: Buffer): void {
+    const hexAddress = this._toHex(address).toLowerCase();
+    const isControlledAddress = this.options.accounts.includes(hexAddress);
+
+    if (!isControlledAddress) {
+      throw new HardhatLedgerNotControlledAddressError(
+        "Tried to send a transaction with an address we don't control.",
+        hexAddress
+      );
+    }
+  }
+
+  private _toHex(value: string | Buffer): string {
```

Comment:
I would add a check here in case an already 0x-prefixed string is passed. Since this function is called several times with many kinds of values, I think it makes sense to be more defensive. Wdyt?

Reply:
Oh good point. I'm having to resist a very strong urge to move it to a `utils.ts` file and properly test it, but I tend to be very careful when adding "bag of methods" files like `utils`.

What do you think? I see a few `utils.ts` and `utilities.ts` in the project but I don't know if you have a preference/way of handling this (besides just leaving it as a private method). For example, I could add a `address.ts` file to `internal` and keep `toHex` there, instead of just going with `utils`

I'll add the check for `0x` locally anyways!

Reply:
Moving it to a separate module to be able to test it is fine by me, yes. My only nitpick is that the file is called `address.ts` but this is used for non-address buffers too :sweat_smile:

Tbh I'm fine with calling it `utils.ts`. But I understand it feels like the lazy option.